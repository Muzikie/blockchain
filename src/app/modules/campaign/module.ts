/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */
import {
	Modules,
	// StateMachine,
} from 'klayr-sdk';
import { CampaignEndpoint } from './endpoint';
import { CampaignMethod } from './method';

export class CampaignModule extends Modules.BaseModule {
	public endpoint = new CampaignEndpoint(this.stores, this.offchainStores);
	public method = new CampaignMethod(this.stores, this.events);
	public commands = [];

	// public constructor() {
	// 	super();
	// 	// registeration of stores and events
	// }

	public metadata(): Modules.ModuleMetadata {
		return {
			...this.baseMetadata(),
			endpoints: [],
			assets: [],
		};
	}

	// Lifecycle hooks
	// public async init(_args: Modules.ModuleInitArgs): Promise<void> {
	// 	// initialize this module when starting a node
	// }

	// public async insertAssets(_context: StateMachine.InsertAssetContext) {
	// 	// initialize block generation, add asset
	// }

	// public async verifyAssets(_context: StateMachine.BlockVerifyContext): Promise<void> {
	// 	// verify block
	// }

	// Lifecycle hooks
	// public async verifyTransaction(_context: StateMachine.TransactionVerifyContext): Promise<StateMachine.VerificationResult> {
	// verify transaction will be called multiple times in the transaction pool
	// return { status: StateMachine.VerifyStatus.OK };
	// }

	// public async beforeCommandExecute(_context: StateMachine.TransactionExecuteContext): Promise<void> {}

	// public async afterCommandExecute(_context: StateMachine.TransactionExecuteContext): Promise<void> {}

	// public async initGenesisState(_context: StateMachine.GenesisBlockExecuteContext): Promise<void> {}

	// public async finalizeGenesisState(_context: StateMachine.GenesisBlockExecuteContext): Promise<void> {}

	// public async beforeTransactionsExecute(_context: StateMachine.BlockExecuteContext): Promise<void> {}

	// public async afterTransactionsExecute(_context: StateMachine.BlockAfterExecuteContext): Promise<void> {
}
