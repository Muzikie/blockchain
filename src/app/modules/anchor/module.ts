/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */

import {
  BaseModule,
  ModuleInitArgs,
  InsertAssetContext,
  BlockVerifyContext,
  TransactionVerifyContext,
  VerificationResult,
  ModuleMetadata,
  TokenMethod,
  // TransactionExecuteContext,
  // GenesisBlockExecuteContext,
  // BlockExecuteContext,
  // BlockAfterExecuteContext,
} from 'lisk-sdk';
// eslint-disable-next-line import/no-extraneous-dependencies
import { VerifyStatus } from 'lisk-framework';
import {
  accountStoreSchema,
  addressRequestSchema,
  idRequestSchema,
  anchorStoreSchema,
  hasAnchorResponse,
} from './schemas';
import { AnchorEndpoint } from './endpoint';
import { AnchorMethod } from './method';
import { AnchorStore } from './stores/anchor';
import { AnchorAccountStore } from './stores/anchorAccount';
import { AnchorCreated } from './events/anchorCreated';
import { CreateCommand } from './commands/create_command';
import { VoteCommand } from './commands/vote_command';

export class AnchorModule extends BaseModule {
  public endpoint = new AnchorEndpoint(this.stores, this.offchainStores);
  public method = new AnchorMethod(this.stores, this.events);

  private readonly _createCommand = new CreateCommand(this.stores, this.events);
  private readonly _purchaseCommand = new VoteCommand(this.stores, this.events);

  // eslint-disable-next-line @typescript-eslint/member-ordering
  public commands = [this._createCommand, this._purchaseCommand];

  private _tokenMethod!: TokenMethod;

  public constructor() {
    super();
    this.stores.register(AnchorAccountStore, new AnchorAccountStore(this.name, 0));
    this.stores.register(AnchorStore, new AnchorStore(this.name, 1));
    this.events.register(AnchorCreated, new AnchorCreated(this.name));
  }

  public addDependencies(tokenMethod: TokenMethod): void {
    this._tokenMethod = tokenMethod;

    this._purchaseCommand.addDependencies(this._tokenMethod);
g  }

  public metadata(): ModuleMetadata {
    return {
      ...this.baseMetadata(),
      endpoints: [
        {
          name: this.endpoint.getAccount.name,
          request: addressRequestSchema,
          response: accountStoreSchema,
        },
        {
          name: this.endpoint.getAnchor.name,
          request: idRequestSchema,
          response: anchorStoreSchema,
        },
        {
          name: this.endpoint.hasAnchor.name,
          request: addressRequestSchema,
          response: hasAnchorResponse,
        },
      ],
      assets: [],
    };
  }

  // Lifecycle hooks
  public async init(_args: ModuleInitArgs): Promise<void> {
    // initialize this module when starting a node
  }

  public async insertAssets(_context: InsertAssetContext) {
    // initialize block generation, add asset
  }

  public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
    // verify block
  }

  // Lifecycle hooks
  // eslint-disable-next-line @typescript-eslint/require-await
  public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
    // verify transaction will be called multiple times in the transaction pool
    return {
      status: VerifyStatus.OK,
    };
  }

  // public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {}

  // public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {}

  // public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {}

  // public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {}
}
