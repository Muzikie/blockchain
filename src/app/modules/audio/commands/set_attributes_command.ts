/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { AudioStore } from '../stores/audio';
import { setAttributesCommandParamsSchema } from '../schemas';
import { SetAttributesCommandParams, Audio } from '../types';
import { validGenres, MIN_RELEASE_YEAR } from '../constants';

export class SetAttributesCommand extends BaseCommand {
  public schema = setAttributesCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(context: CommandVerifyContext<SetAttributesCommandParams>): Promise<VerificationResult> {
    const thisYear = new Date().getFullYear();
    const numericYear = Number(context.params.releaseYear);
    if (numericYear < MIN_RELEASE_YEAR || numericYear > thisYear) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error(`Release year must be a number between ${MIN_RELEASE_YEAR} and ${thisYear}`)
      }
    }
    if (context.params.genre.some(item => item > validGenres.length)) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('Genres should be selected from the list of valid genres')
      }
    }
    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<SetAttributesCommandParams>): Promise<void> {
    const { params, transaction } = context;
    // Get namehash output of the audio file

    const audioSubStore = this.stores.get(AudioStore);

    // Check uniqueness of the NFT
    const audioExists = await audioSubStore.has(context, params.audioID);
    if (!audioExists) {
      throw new Error('Audio with this ID does not exist.');
    }

    const audioNFT: Audio = await audioSubStore.get(context, params.audioID);

    // Check if the sender owns the audio
    if (!audioNFT.creatorAddress.equals(transaction.senderAddress)) {
      throw new Error('You cannot update an audio that you do not own.');
    }

    // Create the Audio object and save it on the blockchain
    const updatedObject: Audio = {
      ...params,
      // set income back
      owners: audioNFT.owners,
      creatorAddress: audioNFT.creatorAddress,
    };
    await audioSubStore.set(context,  params.audioID, updatedObject);
  }
}
