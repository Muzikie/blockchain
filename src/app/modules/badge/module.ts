/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */

import {
  BaseModule,
  ModuleInitArgs,
  InsertAssetContext,
  BlockVerifyContext,
  TransactionVerifyContext,
  VerificationResult,
  ModuleMetadata,
  TransactionExecuteContext,
  // GenesisBlockExecuteContext,
  // BlockExecuteContext,
  // BlockAfterExecuteContext,
} from 'lisk-sdk';
// eslint-disable-next-line import/no-extraneous-dependencies
import { TokenMethod, VerifyStatus } from 'lisk-framework';
import { CreateCommand } from './commands/create_command';
import { DestroyCommand } from './commands/destroy_command';
import { TransferCommand } from './commands/transfer_command';
import { StreamCommand } from './commands/stream_command';
import { ReclaimCommand } from './commands/reclaim_command';
import { SetAttributesCommand } from './commands/set_attributes_command';
import { BadgeCreated } from './events/badgeCreated';
import { BadgeStreamed } from './events/badgeStreamed';
import { BadgeIncomeReclaimed } from './events/badgeIncomeReclaimed';
import { BadgeEndpoint } from './endpoint';
import {
  accountStoreSchema,
  badgeStoreSchema,
  idRequestSchema,
  addressRequestSchema,
} from './schemas';
import { BadgeMethod } from './method';
import { BadgeAccountStore } from './stores/badgeAccount';
import { BadgeStore } from './stores/badge';
import { AnchorMethod } from '../anchor/method';
// import { COMMANDS, MODULES } from '../../constants';

export class BadgeModule extends BaseModule {
public endpoint = new BadgeEndpoint(this.stores, this.offchainStores);
public method = new BadgeMethod(this.stores, this.events);

private readonly _createCommand = new CreateCommand(this.stores, this.events);
private readonly _destroyCommand = new DestroyCommand(this.stores, this.events);
private readonly _transferCommand = new TransferCommand(this.stores, this.events);
private readonly _setAttributesCommand = new SetAttributesCommand(this.stores, this.events);
private readonly _streamCommands = new StreamCommand(this.stores, this.events);
private readonly _reclaimCommands = new ReclaimCommand(this.stores, this.events);

// eslint-disable-next-line @typescript-eslint/member-ordering
public commands = [
  this._createCommand,
  this._destroyCommand,
  this._transferCommand,
  this._setAttributesCommand,
  this._streamCommands,
  this._reclaimCommands,
];

private _anchorMethod!: AnchorMethod;
private _tokenMethod!: TokenMethod;

public constructor() {
  super();
  this.stores.register(BadgeAccountStore, new BadgeAccountStore(this.name, 0));
  this.stores.register(BadgeStore, new BadgeStore(this.name, 1));
  this.events.register(BadgeCreated, new BadgeCreated(this.name));
  this.events.register(BadgeStreamed, new BadgeStreamed(this.name));
  this.events.register(BadgeIncomeReclaimed, new BadgeIncomeReclaimed(this.name));
}

public addDependencies(
  anchorMethod: AnchorMethod,
  tokenMethod: TokenMethod,
): void {
  this._anchorMethod = anchorMethod;
  this._tokenMethod = tokenMethod;

  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  this._streamCommands.addDependencies(this._anchorMethod);
  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  this._reclaimCommands.addDependencies(this._tokenMethod);
}

public metadata(): ModuleMetadata {
  return {
    ...this.baseMetadata(),
    endpoints: [
      {
        name: this.endpoint.getAccount.name,
        request: addressRequestSchema,
        response: accountStoreSchema,
      },
      {
        name: this.endpoint.getBadge.name,
        request: idRequestSchema,
        response: badgeStoreSchema,
      },
    ],
    assets: [],
  };
}

// Lifecycle hooks
public async init(_args: ModuleInitArgs): Promise<void> {
  // initialize this module when starting a node
}

public async insertAssets(_context: InsertAssetContext) {
  // initialize block generation, add asset
}

public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
  // verify block
}

// Lifecycle hooks
// eslint-disable-next-line @typescript-eslint/require-await
public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
  // verify transaction will be called multiple times in the transaction pool
  return {
    status: VerifyStatus.OK,
  };
}

public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {
  // const badgeAccountSubStore = this.stores.get(BadgeAccountStore);
  // const badgeSubStore = this.stores.get(BadgeStore);

//   if (
//     context.transaction.command === COMMANDS.DESTROY &&
// context.transaction.module === MODULES.AUDIO
//   ) {
//     const account = await badgeAccountSubStore.get(context, context.transaction.senderAddress);
//     const badgeID = account.badge.badges[account.badge.badges.length - 1];
//     if (badgeID) {
//       const { collectionID } = await badgeSubStore.get(context, badgeID);
//       // eslint-disable-next-line @typescript-eslint/no-unsafe-call
//       await this._collectionMethod.removeBadge(
//         context.getMethodContext(),
//         badgeID,
//         collectionID,
//         context.transaction.senderAddress,
//       );
//     }
//   }
}

public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {
  // const badgeAccountSubStore = this.stores.get(BadgeAccountStore);
  // const badgeSubStore = this.stores.get(BadgeStore);

  // if (
  //   context.transaction.command === COMMANDS.CREATE &&
  //   context.transaction.module === MODULES.AUDIO
  // ) {
  //   const account = await badgeAccountSubStore.get(context, context.transaction.senderAddress);
  //   const badgeID = account.badge.badges[account.badge.badges.length - 1];
  //   if (badgeID) {
  //     const { collectionID } = await badgeSubStore.get(context, badgeID);
  //     // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  //     await this._collectionMethod.addBadge(
  //       context.getMethodContext(),
  //       badgeID,
  //       collectionID,
  //       context.transaction.senderAddress,
  //     );
  //   }
  // }
}

  // public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {}

  // public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {}
}
