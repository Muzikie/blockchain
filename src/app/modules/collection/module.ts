/* eslint-disable class-methods-use-this */

import {
  BaseModule,
  ModuleInitArgs,
  InsertAssetContext,
  BlockVerifyContext,
  TransactionVerifyContext,
  VerificationResult,
  ModuleMetadata,
  // TransactionExecuteContext,
  // GenesisBlockExecuteContext,
  // BlockExecuteContext,
  // BlockAfterExecuteContext,
} from 'lisk-sdk';
// eslint-disable-next-line import/no-extraneous-dependencies
import { VerifyStatus } from 'lisk-framework';
import { CollectionEndpoint } from './endpoint';
import { CollectionMethod } from './method';
import {
  collectionStoreSchema,
  accountStoreSchema,
  addressRequestSchema,
  idRequestSchema,
} from './schemas';
import { CollectionAccountStore } from './stores/collectionAccount';
import { DestroyCommand } from './commands/destroy_command';
import { TransferCommand } from './commands/transfer_command';
import { SetAttributesCommand } from './commands/set_attributes_command';
import { CollectionStore } from './stores/collection';
import { CreateCommand } from './commands/create_command';
import { CollectionCreated } from './events/collectionCreated'
import { CollectionAttributeSet } from './events/collectionAttributeSet'

export class CollectionModule extends BaseModule {
  public endpoint = new CollectionEndpoint(this.stores, this.offchainStores);
  public method = new CollectionMethod(this.stores, this.events);
  public commands = [
    new CreateCommand(this.stores, this.events),
    new DestroyCommand(this.stores, this.events),
    new TransferCommand(this.stores, this.events),
    new SetAttributesCommand(this.stores, this.events),
  ];

  public constructor() {
    super();
    this.stores.register(CollectionAccountStore, new CollectionAccountStore(this.name, 0));
    this.stores.register(CollectionStore, new CollectionStore(this.name, 1));
    this.events.register(CollectionCreated, new CollectionCreated(this.name));
    this.events.register(CollectionAttributeSet, new CollectionAttributeSet(this.name));
  }

  public metadata(): ModuleMetadata {
    return {
      stores: [],
      endpoints: [
        {
          name: this.endpoint.getAccount.name,
          request: addressRequestSchema,
          response: accountStoreSchema,
        },
        {
          name: this.endpoint.getCollection.name,
          request: idRequestSchema,
          response: collectionStoreSchema,
        },
      ],
      commands: this.commands.map(command => ({
        name: command.name,
        params: command.schema,
      })),
      events: this.events.values().map(v => ({
        name: v.name,
        data: v.schema,
      })),
      assets: [],
    };
  }

  // Lifecycle hooks
  public async init(_args: ModuleInitArgs): Promise<void> {
    // initialize this module when starting a node
  }

  public async insertAssets(_context: InsertAssetContext) {
    // initialize block generation, add asset
  }

  public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
    // verify block
  }

  // Lifecycle hooks
  // eslint-disable-next-line @typescript-eslint/require-await
  public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
    // verify transaction will be called multiple times in the transaction pool
    return {
      status: VerifyStatus.OK,
    };
  }

  // public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {}

  // public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {}

  // public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {}

  // public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {}
}
