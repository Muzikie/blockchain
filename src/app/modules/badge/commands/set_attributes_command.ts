/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { BadgeStore } from '../stores/badge';
import { setAttributesCommandParamsSchema } from '../schemas';
import { SetAttributesCommandParams, Badge } from '../types';
import { VALID_GENRES, MIN_RELEASE_YEAR } from '../constants';

export class SetAttributesCommand extends BaseCommand {
  public schema = setAttributesCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<SetAttributesCommandParams>,
  ): Promise<VerificationResult> {
    const thisYear = new Date().getFullYear();
    const numericYear = Number(context.params.releaseYear);
    if (numericYear < MIN_RELEASE_YEAR || numericYear > thisYear) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error(
          `Release year must be a number between ${MIN_RELEASE_YEAR} and ${thisYear}`,
        ),
      };
    }
    if (context.params.genre.some(item => item > VALID_GENRES.length)) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('Genres should be selected from the list of valid genres'),
      };
    }
    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<SetAttributesCommandParams>): Promise<void> {
    const { params, transaction } = context;
    // Get namehash output of the badge file

    const badgeSubStore = this.stores.get(BadgeStore);

    // Check uniqueness of the NFT
    const badgeExists = await badgeSubStore.has(context, params.badgeID);
    if (!badgeExists) {
      throw new Error('Badge with this ID does not exist.');
    }

    const badgeNFT: Badge = await badgeSubStore.get(context, params.badgeID);

    // Check if the sender owns the badge
    if (!badgeNFT.creatorAddress.equals(transaction.senderAddress)) {
      throw new Error('You cannot update an badge that you do not own.');
    }

    // Create the Badge object and save it on the blockchain
    const updatedObject: Badge = {
      ...params,
      // set income back
      badgeSignature: badgeNFT.badgeSignature,
      badgeHash: badgeNFT.badgeHash,
      owners: badgeNFT.owners,
      creatorAddress: badgeNFT.creatorAddress,
    };
    await badgeSubStore.set(context, params.badgeID, updatedObject);
  }
}
