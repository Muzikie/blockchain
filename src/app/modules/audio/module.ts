/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */

import {
  BaseModule,
  ModuleInitArgs,
  InsertAssetContext,
  BlockVerifyContext,
  TransactionVerifyContext,
  VerificationResult,
  ModuleMetadata,
  TransactionExecuteContext,
  // GenesisBlockExecuteContext,
  // BlockExecuteContext,
  // BlockAfterExecuteContext,
} from 'lisk-sdk';
// eslint-disable-next-line import/no-extraneous-dependencies
import { TokenMethod, VerifyStatus } from 'lisk-framework';
import { CreateCommand } from './commands/create_command';
import { DestroyCommand } from './commands/destroy_command';
import { TransferCommand } from './commands/transfer_command';
import { StreamCommand } from './commands/stream_command';
import { ReclaimCommand } from './commands/reclaim_command';
import { SetAttributesCommand } from './commands/set_attributes_command';
import { AudioCreated } from './events/audioCreated';
import { AudioStreamed } from './events/audioStreamed';
import { AudioIncomeReclaimed } from './events/audioIncomeReclaimed';
import { AudioEndpoint } from './endpoint';
import {
  accountStoreSchema,
  audioStoreSchema,
  idRequestSchema,
  addressRequestSchema,
} from './schemas';
import { AudioMethod } from './method';
import { AudioAccountStore } from './stores/audioAccount';
import { AudioStore } from './stores/audio';
import { CollectionMethod } from '../collection/method';
import { SubscriptionMethod } from '../subscription/method';
import { COMMANDS, MODULES } from '../../constants';

export class AudioModule extends BaseModule {
public endpoint = new AudioEndpoint(this.stores, this.offchainStores);
public method = new AudioMethod(this.stores, this.events);

private readonly _createCommand = new CreateCommand(this.stores, this.events);
private readonly _destroyCommand = new DestroyCommand(this.stores, this.events);
private readonly _transferCommand = new TransferCommand(this.stores, this.events);
private readonly _setAttributesCommand = new SetAttributesCommand(this.stores, this.events);
private readonly _streamCommands = new StreamCommand(this.stores, this.events);
private readonly _reclaimCommands = new ReclaimCommand(this.stores, this.events);

// eslint-disable-next-line @typescript-eslint/member-ordering
public commands = [
  this._createCommand,
  this._destroyCommand,
  this._transferCommand,
  this._setAttributesCommand,
  this._streamCommands,
  this._reclaimCommands,
];

private _collectionMethod!: CollectionMethod;
private _subscriptionMethod!: SubscriptionMethod;
private _tokenMethod!: TokenMethod;

public constructor() {
  super();
  this.stores.register(AudioAccountStore, new AudioAccountStore(this.name, 0));
  this.stores.register(AudioStore, new AudioStore(this.name, 1));
  this.events.register(AudioCreated, new AudioCreated(this.name));
  this.events.register(AudioStreamed, new AudioStreamed(this.name));
  this.events.register(AudioIncomeReclaimed, new AudioIncomeReclaimed(this.name));
}

public addDependencies(
  collectionMethod: CollectionMethod,
  subscriptionMethod: SubscriptionMethod,
  tokenMethod: TokenMethod,
): void {
  this._collectionMethod = collectionMethod;
  this._subscriptionMethod = subscriptionMethod;
  this._tokenMethod = tokenMethod;

  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  this._streamCommands.addDependencies(this._subscriptionMethod);
  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  this._reclaimCommands.addDependencies(this._tokenMethod);
}

public metadata(): ModuleMetadata {
  return {
    ...this.baseMetadata(),
    endpoints: [
      {
        name: this.endpoint.getAccount.name,
        request: addressRequestSchema,
        response: accountStoreSchema,
      },
      {
        name: this.endpoint.getAudio.name,
        request: idRequestSchema,
        response: audioStoreSchema,
      },
    ],
    assets: [],
  };
}

// Lifecycle hooks
public async init(_args: ModuleInitArgs): Promise<void> {
  // initialize this module when starting a node
}

public async insertAssets(_context: InsertAssetContext) {
  // initialize block generation, add asset
}

public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
  // verify block
}

// Lifecycle hooks
// eslint-disable-next-line @typescript-eslint/require-await
public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
  // verify transaction will be called multiple times in the transaction pool
  return {
    status: VerifyStatus.OK,
  };
}

public async beforeCommandExecute(context: TransactionExecuteContext): Promise<void> {
  const audioAccountSubStore = this.stores.get(AudioAccountStore);
  const audioSubStore = this.stores.get(AudioStore);

  if (
    context.transaction.command === COMMANDS.DESTROY &&
context.transaction.module === MODULES.AUDIO
  ) {
    const account = await audioAccountSubStore.get(context, context.transaction.senderAddress);
    const audioID = account.audio.audios[account.audio.audios.length - 1];
    if (audioID) {
      const { collectionID } = await audioSubStore.get(context, audioID);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      await this._collectionMethod.removeAudio(
        context.getMethodContext(),
        audioID,
        collectionID,
        context.transaction.senderAddress,
      );
    }
  }
}

public async afterCommandExecute(context: TransactionExecuteContext): Promise<void> {
  const audioAccountSubStore = this.stores.get(AudioAccountStore);
  const audioSubStore = this.stores.get(AudioStore);

  if (
    context.transaction.command === COMMANDS.CREATE &&
    context.transaction.module === MODULES.AUDIO
  ) {
    const account = await audioAccountSubStore.get(context, context.transaction.senderAddress);
    const audioID = account.audio.audios[account.audio.audios.length - 1];
    if (audioID) {
      const { collectionID } = await audioSubStore.get(context, audioID);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      await this._collectionMethod.addAudio(
        context.getMethodContext(),
        audioID,
        collectionID,
        context.transaction.senderAddress,
      );
    }
  }
}

  // public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {}

  // public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {}
}
