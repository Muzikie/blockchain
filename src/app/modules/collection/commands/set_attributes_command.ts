/* eslint-disable class-methods-use-this */

import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { CollectionStore } from '../stores/collection';
import { setAttributesCommandParamsSchema } from '../schemas';
import { SetAttributesCommandParams, Collection } from '../types';
import { validCollectionTypes, MIN_RELEASE_YEAR } from '../constants';

export class SetAttributesCommand extends BaseCommand {
  public schema = setAttributesCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<SetAttributesCommandParams>,
  ): Promise<VerificationResult> {
    const thisYear = new Date().getFullYear();
    const numericYear = Number(context.params.releaseYear);
    if (numericYear < MIN_RELEASE_YEAR || numericYear > thisYear) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error(
          `Release year must be a number between ${MIN_RELEASE_YEAR} and ${thisYear}`,
        ),
      };
    }
    if (!validCollectionTypes.includes(context.params.collectionType)) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('Type should be selected from the list of valid types'),
      };
    }
    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<SetAttributesCommandParams>): Promise<void> {
    const { params, transaction } = context;
    // Get namehash output of the collection file

    const collectionSubStore = this.stores.get(CollectionStore);

    // Check uniqueness of the NFT
    const collectionExists = await collectionSubStore.has(context, params.collectionID);
    if (!collectionExists) {
      throw new Error('Collection with this ID does not exist.');
    }

    const collectionNFT: Collection = await collectionSubStore.get(context, params.collectionID);

    // Check if the sender owns the collection
    if (!collectionNFT.creatorAddress.equals(transaction.senderAddress)) {
      throw new Error('You cannot update an collection that you do not own.');
    }

    // Create the Collection object and save it on the blockchain
    // Note: You can not change the list of audios using this method
    // Audios of a collection can be changes from the audio module
    const updatedObject: Collection = {
      ...params,
      coverSignature: collectionNFT.coverSignature,
      coverHash: collectionNFT.coverHash,
      audios: collectionNFT.audios,
      creatorAddress: collectionNFT.creatorAddress,
    };
    await collectionSubStore.set(context, params.collectionID, updatedObject);
  }
}
