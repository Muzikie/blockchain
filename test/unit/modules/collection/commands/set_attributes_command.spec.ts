import { SetAttributesCommand } from '../../../../../src/app/modules/collection/commands/set_attributes_command';
import { CollectionModule } from '../../../../../src/app/modules/collection/module';

describe('SetAttributesCommand', () => {
  let command: SetAttributesCommand;
	const module = new CollectionModule();

	beforeEach(() => {
		command = new SetAttributesCommand(module.stores, module.events);
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(command.name).toEqual('setAttributes');
		});

		it('should have valid schema', () => {
			expect(command.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		describe('schema validation', () => {
      it.todo('should throw errors for invalid schema');
      it.todo('should be ok for valid schema');
    });
	});

	describe('execute', () => {
    describe('valid cases', () => {
      it.todo('should update the state store');
    });

    describe('invalid cases', () => {
      it.todo('should throw error');
    });
	});
});
