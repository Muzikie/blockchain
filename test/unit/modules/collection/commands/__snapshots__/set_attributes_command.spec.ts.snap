// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetAttributesCommand constructor should have valid schema 1`] = `
{
  "$id": "collection/setAttributes",
  "properties": {
    "collectionID": {
      "dataType": "bytes",
      "fieldNumber": 4,
    },
    "collectionType": {
      "dataType": "uint32",
      "fieldNumber": 3,
    },
    "name": {
      "dataType": "string",
      "fieldNumber": 1,
      "maxLength": 40,
      "minLength": 3,
    },
    "releaseYear": {
      "dataType": "string",
      "fieldNumber": 2,
    },
  },
  "required": [
    "name",
    "releaseYear",
    "collectionType",
    "collectionID",
  ],
  "title": "SetAttributesAsset transaction asset for collection module",
  "type": "object",
}
`;
